[Login] Improvements to Application Passwords login to work with some Captcha plugins

hichamboushaba opened the pull request:
Closes: #13769 
<!-- Id number of the GitHub issue this PR addresses. -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

### Description
This PR has two main changes:

- It starts web authorization when detecting the word `captcha` in the error message.
- It shows a button to start the web authorization manually for the INVALID_CREDENTIALS error type.

Regarding the second point, we had this button initially, and it was removed during the better error messages project, but this issue that we are fixing here shows that some plugins will use the same error slot for their own error messages, meaning that we can't be sure the issue is always about invalid credentials. To allow the users who might get stuck on these situations, I added back the button to start the web authorization manually.

### Steps to reproduce
**TC 1:**
1. Install this [plugin](https://wordpress.org/plugins/wp-advanced-math-captcha/).
2. Open the app.
3. Sign in using site credentials.

**TC2:**
1. Uninstall the captcha plugin.
2. Open the app.
3. Enter some wrong site credentials.

### Testing information
- For TC1: confirm that the app starts the web authorization flow automatically.
- For TC2: confirm the presence of the button "Try again with WP Admin page"

### The tests that have been performed
^

### Recording
| Before | After |
| --- | --- |
| <video src=https://github.com/user-attachments/assets/95476386-36eb-44a6-badb-9e254e025a4b/> |  <video src=https://github.com/user-attachments/assets/60fb99b3-3016-42c6-9605-4cdf507cdc94/> |

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Remove redundant error type

hichamboushaba committed:
Show actual error message in the dialog

hichamboushaba committed:
Add a button to start web authorization manually

hichamboushaba committed:
Treat error messages that contain `captcha` as invalid

hichamboushaba committed:
Update error message wording

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Add a release note

hichamboushaba commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpapi/Nonce.kt:
```diff
@@ -17,7 +17,6 @@ sealed interface Nonce {
     data class Unknown(override val username: String?) : Nonce
 
     enum class CookieNonceErrorType {
-        NOT_AUTHENTICATED,
```
This error type was redundant with the addition of `INVALID_CREDENTIALS`, so I removed it to simplify the logic

hichamboushaba commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpapi/NonceRestClient.kt:
```diff
@@ -69,10 +69,10 @@ class NonceRestClient @Inject constructor(
 
-                val errorType = if (hasInvalidCredentialsPattern(response.data.orEmpty())) {
+                val errorType = if (hasInvalidCredentialsPattern(response.data.orEmpty()) &&
+                    errorMessage?.contains("captcha", ignoreCase = true) != true
```
This won't work for all cases, and it's another reason why I added back the button "Try again with WP Admin page" as explained in the PR description.

hichamboushaba committed:
Add and fix unit tests

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/WPApiSiteRepository.kt:
```diff
@@ -193,7 +193,7 @@ class WPApiSiteRepository @Inject constructor(
 
     private fun Error.mapToUiString() = when (type) {
-        INVALID_CREDENTIALS -> UiStringRes(string.login_invalid_credentials_message)
+        INVALID_CREDENTIALS -> message?.let { UiStringText(it) } ?: UiStringRes(string.login_invalid_credentials_message)
```
Are you aware that this backend error message is not localized? If you made this change intentionally, that’s fine; otherwise, you might want to consider reverting it.

<img src="https://github.com/user-attachments/assets/fb9fbeb4-156f-445a-9cdc-f5fe0423efdd" width=360>

hichamboushaba replied:
The error message is localized by the site:

<img src=https://github.com/user-attachments/assets/b2cbedba-761e-4cc6-9ca9-b6f18af84ed2 width=360 />

This is not always perfect, as the user could be using a different language than their site, but I think it should be rare, and anyway it's better than showing a wrong error message, WDYT?

irfano replied:
> it's better than showing a wrong error message

I agree! 👍🏻

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsViewModel.kt:
```diff
@@ -142,6 +142,12 @@ class LoginSiteCredentialsViewModel @Inject constructor(
 
+    fun onStartWebAuthorizationClick() {
+        launch {
+            fetchSiteForTutorial()
```
When navigating back from the next screen, the dialog remains visible. We can dismiss the dialog here.

[Screen_recording_20250317_182148.webm](https://github.com/user-attachments/assets/c59e79c0-f004-44ff-8a95-b8b0ad56d035)

hichamboushaba replied:
Good call, implemented in [7b4571e](https://github.com/woocommerce/woocommerce-android/pull/13776/commits/7b4571eb974ff8d59b365b640f08970684b0c420)

irfano commented on WooCommerce/src/main/res/values/strings.xml:
```diff
@@ -356,7 +356,7 @@
     <string name="login_jetpack_installation_continue_magic_link">Or continue using Magic Link</string>
     <string name="login_jetpack_installation_magic_link_failure">An error occurred while fetching your website, please retry!</string>
-    <string name="login_site_credentials_invalid_response">Login failed with an unexpected response from your site. We are working on fixing this issue.</string>
+    <string name="login_site_credentials_invalid_response">Login failed with an unexpected response from your site.</string>
```
Good call! 👍🏻

irfano commented on libs/fluxc-tests/src/test/java/org/wordpress/android/fluxc/network/rest/wpapi/NonceRestClientTest.kt:
```diff
@@ -76,6 +76,7 @@ class NonceRestClientTest {
         val loginResponse = WPAPIResponse.Success(
             """
             <html>
+              <scrip>${NonceRestClient.INVALID_CREDENTIAL_HTML_PATTERN}</scrip>
```
This should ne `<script>`, right? Same typo for the next test case.

hichamboushaba replied:
Yes, fixed in c6165a8. Thanks

irfano reviewed (approved):
Thanks for taking care of this. I added some non-blocking comments and am approving. LGTM! 👍🏻

hichamboushaba committed:
Fix typo

hichamboushaba committed:
Dismiss the error dialog when starting web authorization

hichamboushaba committed:
Merge branch 'trunk' into issue/13769-fix-captcha-login

hichamboushaba commented:
Thanks @irfano for the great review, I addressed your comments, I'll wait for your opinion about this [point](https://github.com/woocommerce/woocommerce-android/pull/13776/#discussion_r1998998192) before merging the PR.
